internal/provider/resource_application.go

```diff
@@ -2,340 +2,536 @@ package provider
 	}
 
-	var parsedConstraints constraints.Value = constraints.Value{}
-	readConstraints := d.Get("constraints").(string)
-	if readConstraints != "" {
-		parsedConstraints, err = constraints.Parse(readConstraints)
+	var parsedConstraints = constraints.Value{}
+	if plan.Constraints.ValueString() != "" {
+		parsedConstraints, err = constraints.Parse(plan.Constraints.ValueString())
 		if err != nil {
-			return diag.FromErr(err)
+			resp.Diagnostics.AddError("Input Error", fmt.Sprintf("Unable to parse constraints, go error: %s", err))
 		}
 	}
-
-	response, err := client.Applications.CreateApplication(&juju.CreateApplicationInput{
-		ApplicationName: name,
-		ModelUUID:       modelUUID,
+	response, err := r.client.Applications.CreateApplication(&juju.CreateApplicationInput{
+		ApplicationName: plan.ApplicationName.ValueString(),
+		ModelUUID:       modelInfo.UUID,
 		CharmName:       charmName,
 		CharmChannel:    channel,
 		CharmRevision:   revision,
 		CharmSeries:     series,
-		Units:           units,
+		Units:           int(plan.UnitCount.ValueInt64()),
 		Config:          configField,
 		Constraints:     parsedConstraints,
-		Trust:           trust,
+		Trust:           plan.Trust.ValueBool(),
 		Expose:          expose,
-		Placement:       placement,
+		Placement:       plan.Placement.ValueString(),
 	})
-
 	if err != nil {
-		return diag.FromErr(err)
-	}
-
-	// These values can be computed, and so set from the response.
-	if err = d.Set("name", response.AppName); err != nil {
-		return diag.FromErr(err)
-	}
-
-	charm["revision"] = response.Revision
-	charm["series"] = response.Series
-	if err = d.Set("charm", []map[string]interface{}{charm}); err != nil {
-		return diag.FromErr(err)
-	}
-
-	id := fmt.Sprintf("%s:%s", modelName, response.AppName)
-	d.SetId(id)
-
-	return resourceApplicationRead(ctx, d, meta)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create application, got error: %s", err))
+		return
+	}
+	r.trace(fmt.Sprintf("create application resource %q", charmName))
+
+	// Faking this plan until the next Read() is done. The value methods used
+	// will put the actual plan values, or default values for the type, e.g.
+	// false and "". Otherwise, there are terraform errors. If we attempt a
+	// ReadApplication, directly after the CreateApplication, application not
+	// found is likely.
+	plan.Constraints = types.StringValue(plan.Constraints.ValueString())
+	plan.Principal = types.BoolValue(plan.Principal.ValueBool())
+	plan.Placement = types.StringValue(plan.Placement.ValueString())
+
+	// Save plan into Terraform state
+	plan.ApplicationName = types.StringValue(response.AppName)
+	planCharm.Revision = types.Int64Value(int64(response.Revision))
+	planCharm.Series = types.StringValue(response.Series)
+	// TODO 01-Aug-2023
+	// Fix this - need the channel return value, the channel requested might
+	// not exist and a different one was used.
+	// Question: should this only be set if the user specified something?
```

Comment:
I don't think so, because then the read and update will say "gotta add the channel", no?

Reply:
The data is being set to State, we wouldn't be changing the plan. Channel has UseStateForUnknown.

I removed the question because it's a computed value, thus something has be to there.